R/type-git.R

```diff
@@ -93,11 +93,76 @@ download_remote_git <- function(resolution, target, target_tree,
   pkgdir <- file.path(target_tree, resolution$package)
   mkdirp(pkgdir)
   async_git_download_repo(url, ref = sha, output = pkgdir)$
+    then(function(files) {
+      check_git_download_missing_files(
+        resolution$remote[[1]],
+        files = files,
+        output = pkgdir
+      )
+    })$
     then(function() {
       "Got"
     })
 }
 
+
+#' Check Downloaded git Repo for Missing Files
+#'
+#' Throws a warning if any files failed to download which are necessary for the
+#' building of the package from source. Will ignore the failure to download
+#' files outside a specified remote sub-directory or omitted using a
+#' `.Rbuildignore` file.
+#'
+#' @param remote A git remote
+#' @param files A named logical vector with names corresponding to relative
+#'   file paths to each file in the git repository, and a logical value
+#'   indicating whether the file was successfully downloaded.
+#' @param output The directory containing downloaded repository files
+#'
+#' @return NULL, function used for its side-effects
+#'
+
+check_git_download_missing_files <- function(remote, files, output) {
+  # subdir consistently trailing in path separator
+  sep <- .Platform$file.sep
+  subdir <- remote$subdir %||% ""
+  if (nchar(subdir) && !endsWith(subdir, sep)) {
+    subdir <- paste0(subdir, sep)
+  }
+
+  # ignore reporting files outside of subdirectory
+  in_subdir <- startsWith(names(files), subdir)
+  files <- files | !in_subdir
+
+  # ignore reporting files that would be ignored using `.Rbuildignore`
+  ignore_path_parts <- c(output, remote$subdir, ".Rbuildignore")
+  ignore_path <- do.call(file.path, as.list(ignore_path_parts))
+  if (file.exists(ignore_path)) {
+    subdir_path <- character(length(files))
+    subdir_path[in_subdir] <- substring(
+      names(files)[in_subdir],
+      nchar(subdir) + 1
+    )
+
+    ignore <- readLines(ignore_path, warn = FALSE)
+    for (pattern in ignore[nzchar(ignore)]) {
+      is_ignored <- grepl(pattern, subdir_path, perl = TRUE, ignore.case = TRUE)
+      files <- files | (in_subdir & is_ignored)
+    }
+  }
+
+  # warn if any files are necessary, but missing
+  if (any(!files)) {
+    throw(pkg_error(
+      "{sum(!files)} necessary file{?s} {?was/were} unable to be fetched",
+      paste0("'", names(files)[!files], "'", collapse = ", "),
+      .class = "pkgdepends_git_fetch"
+    ))
```

Comment:
This was originally a warning, but if files are required for properly building the package it should probably be an error. 

Ideally we could emit a warning and let R continue and fail when it tries to build the package, but R doesn't have any way to know that the files are missing so it would just happily build a faulty package. I think this would be a worse failure mode so I converted it to an error.